File: qabelbox/src/main/java/de/qabel/qabelbox/navigation/MainNavigator.java
package de.qabel.qabelbox.navigation;

import android.app.Fragment;
import android.content.Intent;
import android.util.Log;

import javax.inject.Inject;

import de.qabel.core.config.Contact;
import de.qabel.core.config.Identity;
import de.qabel.desktop.repository.ContactRepository;
import de.qabel.desktop.repository.IdentityRepository;
import de.qabel.desktop.repository.exception.EntityNotFoundException;
import de.qabel.desktop.repository.exception.PersistenceException;
import de.qabel.qabelbox.R;
import de.qabel.qabelbox.activities.CreateAccountActivity;
import de.qabel.qabelbox.activities.MainActivity;
import de.qabel.qabelbox.fragments.AboutLicencesFragment;
import de.qabel.qabelbox.fragments.ContactFragment;
import de.qabel.qabelbox.fragments.HelpMainFragment;
import de.qabel.qabelbox.fragments.IdentitiesFragment;
import de.qabel.qabelbox.ui.views.ChatFragment;
import kotlin.NotImplementedError;

public class MainNavigator implements Navigator {

    public static final String TAG_CONTACT_CHAT_FRAGMENT = "TAG_CONTACT_CHAT_FRAGMENT";
    public static final String TAG_FILES_FRAGMENT = "TAG_FILES_FRAGMENT";
    public static final String TAG_CONTACT_LIST_FRAGMENT = "TAG_CONTACT_LIST_FRAGMENT";
    public static final String TAG_ABOUT_FRAGMENT = "TAG_ABOUT_FRAGMENT";
    public static final String TAG_HELP_FRAGMENT = "TAG_HELP_FRAGMENT";
    public static final String TAG_MANAGE_IDENTITIES_FRAGMENT = "TAG_MANAGE_IDENTITIES_FRAGMENT";
    public static final String TAG_FILES_SHARE_INTO_APP_FRAGMENT = "TAG_FILES_SHARE_INTO_APP_FRAGMENT";

    private static final String TAG = "MainNavigator";

    @Inject
    MainActivity activity;

    @Inject
    IdentityRepository identityRepository;

    @Inject
    ContactRepository contactRepository;

    @Inject
    Identity activeIdentity;

    @Inject
    public MainNavigator(MainActivity activity,
                         IdentityRepository identityRepository,
                         ContactRepository contactRepository,
                         Identity activeIdentity) {
        this.activity = activity;
        this.identityRepository = identityRepository;
        this.contactRepository = contactRepository;
        this.activeIdentity = activeIdentity;
    }

    @Override
    public void selectCreateAccountActivity() {
        Intent intent = new Intent(activity, CreateAccountActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK | Intent.FLAG_ACTIVITY_TASK_ON_HOME);
        activity.startActivity(intent);
        activity.finish();
    }

    /*
        FRAGMENT SELECTION METHODS
    */
    @Override
    public void selectManageIdentitiesFragment() {
        try {
            showMainFragment(IdentitiesFragment.newInstance(identityRepository.findAll()),
                    TAG_MANAGE_IDENTITIES_FRAGMENT);
        } catch (PersistenceException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void selectChatFragment(String activeContact) {
        if (activeContact == null) {
            selectChatFragment();
            return;
        }
        try {
            Contact contact = contactRepository.findByKeyId(activeIdentity, activeContact);
            Log.d(TAG, "Selecting chat with  contact " + contact.getAlias());
            activity.getFragmentManager().beginTransaction().add(R.id.fragment_container,
                    ChatFragment.Companion.withContact(contact),
                    TAG_CONTACT_CHAT_FRAGMENT)
                    .addToBackStack(TAG_CONTACT_CHAT_FRAGMENT).commit();
        } catch (EntityNotFoundException entityNotFoundException) {
            Log.w(TAG, "Could not find contact " + activeContact);
            selectChatFragment();
        }
    }

    @Override
    public void selectChatFragment() {
        showMainFragment(new ContactFragment(), TAG_CONTACT_LIST_FRAGMENT);
    }

    @Override
    public void selectHelpFragment() {
        showMainFragment(new HelpMainFragment(), TAG_HELP_FRAGMENT);
    }

    @Override
    public void selectAboutFragment() {
        showMainFragment(AboutLicencesFragment.newInstance(), TAG_ABOUT_FRAGMENT);
    }

    @Override
    public void selectFilesFragment() {
        throw new NotImplementedError();
    }

    private void showMainFragment(Fragment fragment, String tag) {
        activity.getFragmentManager().beginTransaction()
                .replace(R.id.fragment_container, fragment, tag).commit();
        try {
            while (activity.getFragmentManager().executePendingTransactions()) {
                Thread.sleep(50);
            }
        } catch (InterruptedException e) {
            Log.e(TAG, "Error waiting for fragment change", e);
        }
        activity.handleMainFragmentChange();
    }


}

File: qabelbox/src/main/java/de/qabel/qabelbox/navigation/MainNavigator.kt
package de.qabel.qabelbox.navigation

import android.app.Fragment
import android.content.Context
import android.content.Intent
import de.qabel.core.config.Contact
import de.qabel.core.config.Identity
import de.qabel.core.repository.ContactRepository
import de.qabel.core.repository.IdentityRepository
import de.qabel.core.repository.exception.EntityNotFoundException
import de.qabel.core.repository.exception.PersistenceException
import de.qabel.qabelbox.R
import de.qabel.qabelbox.activities.CreateAccountActivity
import de.qabel.qabelbox.activities.MainActivity
import de.qabel.qabelbox.box.views.FileBrowserFragment
import de.qabel.qabelbox.contacts.dto.ContactDto
import de.qabel.qabelbox.contacts.view.views.ContactDetailsFragment
import de.qabel.qabelbox.contacts.view.views.ContactsFragment
import de.qabel.qabelbox.fragments.AboutLicencesFragment
import de.qabel.qabelbox.fragments.HelpMainFragment
import de.qabel.qabelbox.fragments.IdentitiesFragment
import de.qabel.qabelbox.fragments.QRcodeFragment
import de.qabel.qabelbox.chat.view.views.ChatFragment
import de.qabel.qabelbox.chat.view.views.ChatOverviewFragment
import de.qabel.qabelbox.contacts.view.views.ContactEditFragment
import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.warn
import javax.inject.Inject

class MainNavigator
@Inject
constructor(var activity: MainActivity,
            var identityRepository: IdentityRepository,
            var contactRepository: ContactRepository,
            var activeIdentity: Identity) : AbstractNavigator(), Navigator, AnkoLogger {

    companion object {
        const val TAG_CONTACT_LIST_FRAGMENT = "TAG_CONTACT_LIST_FRAGMENT"
        const val TAG_CONTACT_CHAT_FRAGMENT = "TAG_CONTACT_CHAT_FRAGMENT"
        const val TAG_CHAT_OVERVIEW_FRAGMENT = "TAG_CHAT_OVERVIEW_FRAGMENT"
        const val TAG_CONTACT_DETAILS_FRAGMENT = "TAG_CONTACT_DETAILS_FRAGMENT";
        const val TAG_CONTACT_EDIT_FRAGMENT = "TAG_CONTACT_EDIT_FRAGMENT";
        const val TAG_QR_CODE_FRAGMENT = "TAG_CONTACT_QR_CODE_FRAGMENT";

        const val TAG_FILES_FRAGMENT = "TAG_FILES_FRAGMENT"
        const val TAG_ABOUT_FRAGMENT = "TAG_ABOUT_FRAGMENT"
        const val TAG_HELP_FRAGMENT = "TAG_HELP_FRAGMENT"
        const val TAG_MANAGE_IDENTITIES_FRAGMENT = "TAG_MANAGE_IDENTITIES_FRAGMENT"
        const val TAG_FILES_SHARE_INTO_APP_FRAGMENT = "TAG_FILES_SHARE_INTO_APP_FRAGMENT"

        fun createChatIntent(context: Context, identityKey: String, contactKey: String) =
                Intent(context, MainActivity::class.java).apply {
                    putExtra(MainActivity.START_CHAT_FRAGMENT, true)
                    putExtra(MainActivity.ACTIVE_IDENTITY, identityKey)
                    putExtra(MainActivity.ACTIVE_CONTACT, contactKey)
                }

    }

    private fun showMainFragment(fragment: Fragment, tag: String) {
        showFragment(activity, fragment, tag, false, true)
    }

    override fun selectCreateAccountActivity() {
        val intent = Intent(activity, CreateAccountActivity::class.java)
        intent.flags = Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_TASK_ON_HOME
        activity.startActivity(intent)
        activity.finish()
    }

    /*
        FRAGMENT SELECTION METHODS
    */
    override fun selectManageIdentitiesFragment() {
        try {
            showMainFragment(IdentitiesFragment.newInstance(identityRepository.findAll()),
                    TAG_MANAGE_IDENTITIES_FRAGMENT)
        } catch (e: PersistenceException) {
            throw RuntimeException(e)
        }

    }

    override fun selectHelpFragment() {
        showMainFragment(HelpMainFragment(), TAG_HELP_FRAGMENT)
    }

    override fun selectAboutFragment() {
        showMainFragment(AboutLicencesFragment(), TAG_ABOUT_FRAGMENT)
    }

    override fun selectFilesFragment()
            = showMainFragment(FileBrowserFragment.newInstance(), TAG_FILES_FRAGMENT)

    override fun selectContactsFragment() {
        showMainFragment(ContactsFragment(), TAG_CONTACT_LIST_FRAGMENT)
    }

    override fun selectChatOverviewFragment() {
        showMainFragment(ChatOverviewFragment(), TAG_CHAT_OVERVIEW_FRAGMENT)
    }

    override fun selectQrCodeFragment(contact: Contact) {
        showFragment(activity, QRcodeFragment.newInstance(contact), TAG_QR_CODE_FRAGMENT, true, false)
    }

    override fun selectContactDetailsFragment(contact: Contact) {
        showFragment(activity, ContactDetailsFragment.withContactKey(contact.keyIdentifier), TAG_CONTACT_DETAILS_FRAGMENT, true, false)
    }

    override fun selectContactDetailsFragment(contactDto: ContactDto) {
        showFragment(activity, ContactDetailsFragment.withContactKey(contactDto.contact.keyIdentifier), TAG_CONTACT_DETAILS_FRAGMENT, true, false)
    }

    override fun selectChatFragment(activeContact: String?) {
        if (activeContact == null) {
            return
        }
        selectContactChat(activeContact, activeIdentity)
    }

    override fun selectContactChat(contactKey: String, withIdentity: Identity) {
        if (activeIdentity.keyIdentifier.equals(withIdentity.keyIdentifier)) {
            try {
                val contact = contactRepository.findByKeyId(withIdentity, contactKey);
                showFragment(activity, ChatFragment.withContact(contact), TAG_CONTACT_CHAT_FRAGMENT, true, true)
            } catch (entityNotFoundException: EntityNotFoundException) {
                warn("Could not find contact " + contactKey, entityNotFoundException)
            }
        } else {
            val intent = createChatIntent(activity, withIdentity.keyIdentifier, contactKey)
            activity.finish()
            activity.overridePendingTransition(R.anim.fade_in, R.anim.fade_out)
            activity.startActivity(intent)
        }
    }

    override fun selectContactEdit(contactDto: ContactDto) {
        showFragment(activity, ContactEditFragment.withContact(contactDto), TAG_CONTACT_EDIT_FRAGMENT, true, false)
    }

    override fun popBackStack() {
        activity.fragmentManager.popBackStack()
    }

}
